### ISSUE TITLE
Add a DFS iterator to Graph

### ISSUE THREAD

**Issue by eric7237cire** [EbTech/rust-algorithms#5]
Hello,

Are you taking pull requests?  I wrote a dfs method that returns an iterator for your graph class.  I'm new to rust so I can't say if its the absolute best.  It is iterative and keeps the state in the iterator struct.  And of course a couple tests.

I also used a bit vector because why not :)  

Let me know if this can be of interest and I'll create a proper PR.  Otherwise, feel free to close this.  Regardless a great repo, thank you !

```
use super::graph::Graph;
use bit_vec::BitVec;

impl Graph
{
    fn dfs(&self, v: usize) -> DfsIterator
    {
        // Create a stack for DFS
        let mut stack: Vec<usize> = Vec::new();

        // Push the current source node.
        stack.push(v);

        DfsIterator {
            graph: self,
            visited: BitVec::from_elem(self.num_v(), false),
            stack,
        }
    }
}
pub struct DfsIterator<'a>
{
    graph: &'a Graph,
    //is vertex visited
    visited: BitVec,
    //stack of vertices
    stack: Vec<usize>,
}

impl<'a> Iterator for DfsIterator<'a>
{
    type Item = usize;

    
    /// Returns next vertex in the DFS
    fn next(&mut self) -> Option<Self::Item>
    {
        let mut r = None;

        //Code translated/adapted from https://www.geeksforgeeks.org/iterative-depth-first-traversal/        
        while !self.stack.is_empty() {
            // Pop a vertex from stack and print it
            let s = self.stack.pop().unwrap();

            // Stack may contain same vertex twice. So
            // we need to print the popped item only
            // if it is not visited.
            if !self.visited[s] {
                self.visited.set(s, true);
                r = Some(s);
            }

            // Get all adjacent vertices of the popped vertex s
            // If a adjacent has not been visited, then puah it
            // to the stack.
            for (_e, u) in self.graph.adj_list(s) {
                if !self.visited[u] {
                    self.stack.push(u);
                }
            }

            if r != None {
                return r;
            }
        }

        return None;
    }
}

#[cfg(test)]
mod test
{
    use super::*;

    #[test]
    fn test_dfs()
    {
        let mut graph = Graph::new(4, 8);
        graph.add_edge(0, 2);
        graph.add_edge(2, 0);
        graph.add_edge(1, 2);
        graph.add_edge(0, 1);
        graph.add_edge(3, 3);
        graph.add_edge(2, 3);

        //start at 2;  -- 2 0 1 3

        let dfs_search = graph.dfs(2).collect::<Vec<_>>();
        assert_eq!(dfs_search, vec![2, 0, 1, 3]);

        ////https://www.geeksforgeeks.org/iterative-depth-first-traversal/
    }

    #[test]
    fn test_dfs2()
    {
        let mut graph = Graph::new(5, 8);
        graph.add_edge(0, 2);
        graph.add_edge(2, 1);
        graph.add_edge(1, 0);
        graph.add_edge(0, 3);
        graph.add_edge(3, 4);
        graph.add_edge(4, 0);

        let dfs_search = graph.dfs(0).collect::<Vec<_>>();
        assert_eq!(dfs_search, vec![0, 2, 1, 3, 4]);
        //0 3 4 2 1 or
        //0 2 1 3 4
    }
}

```

**Comment by EbTech** [comment]
Oh, very nice! This repo is optimized for users who want to create small stand-alone snippets to solve algorithmic problems (e.g., programming contests). If you make dfs() a function instead of a method, it might be easier for users to pick and choose the pieces they need. Please let me know your thoughts.

I haven't thought much about how to organize contributions. Feel free to make a PR, if you're ok with stuff being moved around later :D

Thanks for the contribution!

**Comment by EbTech** [comment]
The same argument would apply to some of my existing functions, like the Euler tours. OK I'll have to find a way to not let this blow up if we continue adding stuff!

So far, I've only added things that I found very general and hard to produce on the spot. At the time, I considered standard DFS/BFS "easy", and not sufficiently general since a lot of problems involve modifying the DFS code anyway. But here you're using the language in a nice way, so it serves as a useful example for anyone wanting to try writing algorithms in Rust!
